### PR TITLE
[ocaml] Update required OCaml version to 4.05.0

### LINKED ISSUE #7380: Minimal OCaml version for Coq 8.10
After Ubuntu Bionic has been released, and after quite a few versions requiring OCaml 4.02.3, it seems to me that it is time for us to raise the lower bound in our OCaml version to 4.05.0.

At the time 8.9 is released, few people should still be in a distro shipping 4.02.3, and those that are will be better served without a doubt by using OPAM. Moving to 4.05.0 will allow us to clean quite a bit of stuff. Debian Buster should also ship with Coq 8.9 and OCaml 4.05.0.

Most of the infrastructure is ready, once we agree, we will push the correspoding changes.

### PR DESCRIPTION
**Pull request by ejgallego**
Closes #7380. Ubuntu 18.04 and Debian Buster will ship this OCaml
version so it makes sense we bump our dependency to 4.05.0 as we can
use some newer compiler features.

### CODE REVIEW

**Review by gares — CHANGES_REQUESTED** [review]
~Change in test suite seems unrelated~ (too early in the morning)

**Review by gares — APPROVED** [review]

**Review by gares — CHANGES_REQUESTED** [review]
You need to fix `dev/ci/appveyor.sh` as well since it contains `opam init -a mingw https://github.com/fdopen/opam-repository-mingw.git --comp 4.02.3+mingw64c --switch 4.02.3+mingw64c`

**Review by MSoegtropIMC — APPROVED** [review]
I just reviewed the changes to the files in the area I own, that is:

dev/build/windows/makecoq_mingw.sh

These changes are fine

**Comment by ejgallego** [comment]
> You need to fix dev/ci/appveyor.sh as well since it contains

Fixed;



> I just reviewed the changes to the files in the area I own, that is:
> 
> dev/build/windows/makecoq_mingw.sh

@MSoegtropIMC I see you are installing a pretty old menhir version. I was tempted to update it, but on the other hand, menhir is unused by the build, so why it is there in the first place?

**Review by gares — APPROVED** [review]

**Comment by MSoegtropIMC** [comment]
@ejgallego: this is a historical artifact. The scripts originate from scripts I used to setup my own environment. If menhir would have any dependencies, this would be documented in the script by calling "make_menhir" somewhere, but it is only called from the top level build, so it is a "top level" feature.

If you like please go ahead and remove menhir, but please leave the make_menhir function in, just remove the call to it. I will add an option for it later.

Btw.: another of these artifacts is "make_stdint". The call to it can also be removed, but again, please keep the function in.

File: .circleci/config.yml
Comment (SkySkimmer) [inline .circleci/config.yml]: That seems like it should be its own PR.

Comment (SkySkimmer) [reply .circleci/config.yml]: Or just remove circle.

**Comment by ejgallego** [comment]
> Btw.: another of these artifacts is "make_stdint". The call to it can also be removed, but again, please keep the function in.

I guess that would speed up the coq build a bit, so will try.

File: .circleci/config.yml
Comment (ejgallego) [reply .circleci/config.yml]: Moved to its own PR.

File: test-suite/coqchk/cumulativity.v
Comment (gares) [reply test-suite/coqchk/cumulativity.v]: I'm curious too ;-)

Comment (ejgallego) [reply test-suite/coqchk/cumulativity.v]: More interesting is the test-suite failure in 4.05.0

**Comment by gares** [comment]
The windows32/64 jobs on gitlab fail here
  https://github.com/coq/coq/blob/master/dev/build/windows/makecoq_mingw.sh#L834
It seems the file is now called `INSTALL.adoc`.
I've no clue why it worked on appveyor ...

**Comment by MSoegtropIMC** [comment]
@gares : As far as I can tell on appveyor only the opam variant did run.

**Comment by MSoegtropIMC** [comment]
P.S.: I would prefer if this would be fixed such that either filename works, so that one can go back to OCaml 4.02 in case one needs this for whatever reason.

**Comment by gares** [comment]
Why do we need to install the "documentation" in a special folder in the first place? I see OCaml does not install it anymore, but is there any need for it?

BTW, the `rm -f ./*txt` seems wrong to me, I guess you wanted to remove txt files in the target directory, not in `.`.

**Comment by MSoegtropIMC** [comment]
@gares : I chose to do so so that users would find all lincecse and readme files in one place rather than distributed all over the installation - probably something industry people are more interested in than research folks ...

The rm -f ./*txt looks indeed bogus and doesn't match the comment string.
I guess it can be removed since it was for compatibility of the build script between 4.01 and 4.02. Since we are moving to 4.05, this doesn't seem to be necessary.

I am anyway working on the scripts and can fix this on the way.

**Comment by ejgallego** [comment]
> I am anyway working on the scripts and can fix this on the way.

That would be great, thanks. Note that for OPAM-built packages documentation is put in a standard place, so usually you can access all of it easily.

**Comment by ejgallego** [comment]
So the test suite failure is due to `camlp5` 7.00 being buggy, upping to 7.01

File: dev/build/windows/makecoq_mingw.sh
Comment (MSoegtropIMC) [inline dev/build/windows/makecoq_mingw.sh]: The file in the chache and the log files will ne names rel705 now.
You should add an optional module name like camlp5-rel705 after the last argument (the 1) if the file name is not a good module name.

**Comment by MSoegtropIMC** [comment]
@ejgallego : P.s.: sorry that I am not that helpful these days - got a new laptop and although it is in principle quite fast, building Coq with it is slow like hell for a reason I couldn't track down as yet even with ProcMon & friends. I need to setup a reasonable working environment for the workshop the week after next week.

File: dev/build/windows/makecoq_mingw.sh
Comment (ejgallego) [reply dev/build/windows/makecoq_mingw.sh]: Done, thanks!

**Comment by ejgallego** [comment]
> @ejgallego : P.s.: sorry that I am not that helpful these days - got a new laptop and although it is in principle quite fast, building Coq with it is slow like hell for a reason I couldn't track down as yet even with ProcMon & friends. I need to setup a reasonable working environment for the workshop the week after next week.

Oh no worries and thanks, this is not very urgent at all, as likely this change will be indeed shipped in 8.10 given the release schedule; I also got a quite fancy new laptop but it has given me a large share of problems :) including the 8650U downclocking itself to 200 MHz when building Coq, but the worse is that this pricey laptop just refuses to POST anymore and Lenovo tells me that it will take months to get repaired... %$!%$

**Review by maximedenes — APPROVED** [review]
I haven't reviewed the patch, but I'm ok with the change in principle.

**Comment by SkySkimmer** [comment]
Who's in charge of this PR?

**Comment by MSoegtropIMC** [comment]
@ejgallego : what about  #7596? I am a bit confused that there are two PRs which do on the WIndows side similar things. #7596 upgrades OCaml on Windows to 4.06.1 and is approved by @maximedenes , @Zimmi48 (and me).

I was about to merge it. Shoudln't I?

**Comment by Zimmi48** [comment]
You should. The present PR needs to be discussed because it changes the minimal supported OCaml version. The Windows-only PR updates the version that is used in practice and require no such discussion.

**Comment by ejgallego** [comment]
Thanks @Zimmi48 , indeed the other PR is just about having a windows build with a (hopefully) more correct runtime for Windows; given that the final packages for the users are built using the CI it makes sense.

**Comment by MSoegtropIMC** [comment]
@Zimmi48, @ejgallego : thanks for the clarification. I just thought this requires a double check. I will do the merge today.

**Review by ppedrot — APPROVED** [review]
No opposition on my side.

**Comment by gares** [comment]
ping @letouzey @mattam82

**Comment by ejgallego** [comment]
I'm gonna try to advance with this today. We are missing @SkySkimmer review in the name of the CI team.

File: dev/ci/docker/bionic_coq/Dockerfile
Comment (SkySkimmer) [inline dev/ci/docker/bionic_coq/Dockerfile]: Why move odoc to BASE_OPAM? Its users haven't changed AFAICS.

**Comment by ejgallego** [comment]
Link to the coq-club thread https://sympa.inria.fr/sympa/arc/coq-club/2018-09/msg00041.html

I do propose we merge this at the start of next week, unless some serious problem happens.

Who wants to be the assignee? A member of the CI team could be a good one.

File: dev/ci/docker/bionic_coq/Dockerfile
Comment (ejgallego) [reply dev/ci/docker/bionic_coq/Dockerfile]: `odoc` is ocaml >= 4.03 [see https://opam.ocaml.org/packages/odoc/] IMHO it makes sense to have it at the base as we may want to build the API docs in all settings.

Anyways it can be kept like that. Dune release still requires >= 4.06

**Comment by SkySkimmer** [comment]
It should be just as easy to move it once there is a need to I think.

Gaëtan Gilbert

**Comment by ejgallego** [comment]
> It should be just as easy to move it once there is a need to I think.

Yeah, the thing is that the BASE_EDGE variable didn't exists a couple of week ago, so I saw it as spurious. I do think it makes sense to have a single OPAM BASE if possible, as to enable uniform testing, but yeah, that's a nit indeed.

**Review by SkySkimmer — APPROVED** [review]
>Who wants to be the assignee? A member of the CI team could be a good one.

Isn't this more a build team kind of thing?

File: .gitlab-ci.yml
Comment (SkySkimmer) [reply .gitlab-ci.yml]: I'm putting needs: progress to remind people not to merge, remove it once you remove the hack.

**Comment by ejgallego** [comment]
> Isn't this more a build team kind of thing?

Yeah, I dunno; maybe the 8.10 release manager should be the one? Anyways I think most developers are OK with the change as basically this has been open for 6 months with no opposition, so anyone could take care of this IMHO.

File: .gitlab-ci.yml
Comment (Zimmi48) [reply .gitlab-ci.yml]: You could open a feature request on the bot issue tracker to automatically run the build with the right `SKIP_DOCKER` value depending on what changed.

**Comment by ejgallego** [comment]
Personally, I consider the moment has finally arrived.

**Comment by SkySkimmer** [comment]
>the 8.10 release manager

Who's that again?

Gaëtan Gilbert

**Comment by Zimmi48** [comment]
No one for now. We didn't have a WG in a really long while BTW.

**Comment by ejgallego** [comment]
This has the support of a good number of developers already, please @herbelin @mattam82 @silene @Zimmi48 @ybertot see if you have any issue with the change, otherwise we will remove the `needs: discussion` tag soon.

**Review by Zimmi48 — APPROVED** [review]

**Comment by herbelin** [comment]
Looks good to me (and very nice mail to coq-club btw).

**Comment by ejgallego** [comment]
Some more data for people curious about dependencies: https://repology.org/metapackage/ocaml/versions

**Comment by ybertot** [comment]
Sorry, I don't have an opinion and don't understand why I am required to give one.

**Comment by Zimmi48** [comment]
@ybertot Emilio just checked with all the members of the core team to avoid waiting for a decision to happen at a WG.

**Comment by ejgallego** [comment]
I feel like we have enough consensus to remove the `needs: discussion` label and to look for the assignee.

File: INSTALL
Comment (ejgallego) [reply INSTALL]: Hihi, I liked the legacy name, it was a kind of easter egg. But indeed, no problem in updating the name.

**Comment by ejgallego** [comment]
Removing the `needs: discussion` tag, this is ready to merge. Natural assignee based on files would be @SkySkimmer , however any other core developer feel free to take this.

**Comment by ppedrot** [comment]
It seems @SkySkimmer lost track of this PR, so I'll merge it instead.

**Comment by Zimmi48** [comment]
The reason for delaying was to wait for 8.9+beta1 to be released. But well, as long as people do not use 4.05 features in PRs that are put in 8.9 milestones, this shouldn't cause any trouble.

**Comment by ppedrot** [comment]
Scheiße, I forgot that we indeed delayed that one... It was not written down in the PR though. Hopefully you're right.

**Comment by ejgallego** [comment]
Indeed we were waiting for the beta as a bit of courtesy, but I guess we are close enough as for this not to matter. By the way, I received some replies in private to my coq-club mail supporting the move to 4.05.0.

**Pull request merged**
